src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/RoslynCSharpTokenizer.cs

```diff
@@ -1,173 +1,101 @@
-                { '}', () => SyntaxKind.RightBrace },
-                { '[', () => SyntaxKind.LeftBracket },
-                { ']', () => SyntaxKind.RightBracket },
-                { ',', () => SyntaxKind.Comma },
-                { ';', () => SyntaxKind.Semicolon },
-                { '~', () => SyntaxKind.Tilde },
-                { '#', () => SyntaxKind.Hash }
-            };
+        _roslynTokenParser = CodeAnalysis.CSharp.SyntaxFactory.CreateTokenParser(source.SourceText, null);
     }
 
-    protected override int StartState => (int)CSharpTokenizerState.Data;
+    protected override int StartState => (int)RoslynCSharpTokenizerState.Start;
 
-    private new CSharpTokenizerState? CurrentState => (CSharpTokenizerState?)base.CurrentState;
+    private new RoslynCSharpTokenizerState? CurrentState => (RoslynCSharpTokenizerState?)base.CurrentState;
 
     public override SyntaxKind RazorCommentKind => SyntaxKind.RazorCommentLiteral;
 
     public override SyntaxKind RazorCommentTransitionKind => SyntaxKind.RazorCommentTransition;
 
     public override SyntaxKind RazorCommentStarKind => SyntaxKind.RazorCommentStar;
 
+    internal override void StartingBlock()
+    {
+        _roslynTokenParser.SkipForwardTo(Source.Position);
+    }
+
+    internal override void EndingBlock()
+    {
+        // We should always be transitioning to the other parser in response to content. This means that we'll have parsed a token, and be in the TriviaForCSharpToken
+        // state. The other possibility is that the parser looked at the current token, and put it back to let the other parser handle it; in this case, we should be in
+        // the Start state. In order to ensure that we properly handle the trailing trivia (because the other parser will handle the trailing trivia on the node we found,
+        // if any), we need to reset back before the start of that node, skip the content, and reset our state back to Start for when we're called back next.
+        if (CurrentState == RoslynCSharpTokenizerState.Start)
+        {
+            return;
+        }
+
+        Debug.Assert(CurrentState == RoslynCSharpTokenizerState.TriviaForCSharpToken, $"Unexpected state: {CurrentState}");
+        Debug.Assert(_currentCSharpTokenTriviaEnumerator is (_, isLeading: false));
+        Debug.Assert(_resultCache.Count > 0);
+
+        var (_, result) = _resultCache[^1];
+        _roslynTokenParser.ResetTo(result);
+        _resultCache.RemoveAt(_resultCache.Count - 1);
+        var lastToken = result.Token;
+        if (lastToken.HasLeadingTrivia)
+        {
+            // If the previous token did indeed have leading trivia, we need to make sure to take it into account so that any preprocessor directives are seen by the
+            // roslyn token parser
+            _ = GetNextResult(NextResultType.LeadingTrivia);
+        }
+
+        _roslynTokenParser.SkipForwardTo(lastToken.Span.End);
+        base.CurrentState = (int)RoslynCSharpTokenizerState.Start;
```

Comment:
Consider adding a setter to the `private new RoslynCSharpTokenizerState? CurrentState` property and using that instead.